# Introduction to EIG

The Eigenstrat format is a text-based format used for storing genotype data in genetics. It consists of three files:

**1. .snp file**: containing the SNP positions. It consists of six columns: SNP-name, chromosome, genetic positions, physical position, reference allele, alternative allele.

**2. .ind file**: containing the names of the individuals. It consists of three columns: Individual Name, Sex (encoded as M(ale), F(emale), or U(nknown)), and population name.

**3. .geno file**: containing the genotype matrix, with individuals laid out from left to right, and SNP positions laid out from top to bottom.
-   0 means zero copies of reference allele.
-   1 means one copy of reference allele.
-   2 means two copies of reference allele.
-   9 means missing data.

MORE INFO: https://reich.hms.harvard.edu/software/InputFileFormats

## Merge datasets in EIGENSTRAT
You will need to install eigensoft as a conda environment:
```
conda activate
conda create -n eigensoft8 
conda install bioconda::eigensoft #v8 if you want to compute PCA with ellipses
```
Use the mergeit function:
```
conda activate eigensoft8

in1=$1
in2=$2
out=$3

mergeit -p <(echo "geno1: ${in1}.geno
snp1:  ${in1}.snp
ind1:  ${in1}.ind
geno2: ${in2}.geno
snp2:  ${in2}.snp
ind2:  ${in2}.ind
genooutfilename:   ${out}.geno
snpoutfilename:    ${out}.snp
indoutfilename:    ${out}.ind
outputformat:  EIGENSTRAT
docheck: YES
strandcheck:  YES
hashcheck: NO")
```
More info: https://github.com/argriffing/eigensoft/blob/master/CONVERTF/README

## Subset in EIGENSTRAT
Prepare a list of individuals/pop labels that you wan tto subset. The labels in that list have to match the labels of the third column in the .ind file.

Use the convertf function:
```
conda activate eigensoft

in1=$1 #geno and snp files prefix
in2=$2 #ind file
out=$3 #output prefix
poplist=$4 #list of ind/pops to subset

convertf -p <(echo "genotypename:	${in1}.geno
snpname:	${in1}.snp
indivname:	${in2}
outputformat:	EIGENSTRAT 
genotypeoutname:	${out}.geno
snpoutname:	${out}.snp
indivoutname:	${out}.ind
poplistname:	${poplist}")
```
More info: https://github.com/argriffing/eigensoft/blob/master/CONVERTF/README

## Convert between PLINK and EIGENSTRAT
The binary PLINK format is a compact and efficient file format used in genetics to store genotype data. It consists of three files:

1. BED file – contains the actual genotype data in binary format.
2. BIM file – contains variant (SNP) information in 6 columns:
  - Chromosome
  - SNP ID
  - Genetic distance (cM) - often 0 if unknown
  - Genomic coordinate
  - Allele 1 (minor allele)
  - Allele 2 (major allele)
3. FAM file – contains sample (individual) information in 6 columns:
  - Family ID
  - Individual ID
  - Paternal ID (0 if missing)
  - Maternal ID (0 if missing)
  - Sex (1=male,2=female,0=unknown)
  - Phenotype: Usually -9/9 (as missing) for popgen analysis


### From EIG to PLINK:
```
conda activate eigensoft

in1=$1

convertf -p <(echo "genotypename:       ${in1}.geno
snpname:        ${in1}.snp
indivname:      ${in1}.ind
outputformat:   PACKEDPED
genotypeoutname:        ${in1}.bed
snpoutname:     ${in1}.bim
indivoutname:   ${in1}.fam")
```

### From PLINK to EIG:
```
conda activate eigensoft
in1=$1

convertf -p <(echo "genotypename:    ${in1}.bed
snpname:         ${in1}.bim
indivname:       ${in1}.fam
outputformat:    EIGENSTRAT
genotypeoutname: ${in1}.geno
snpoutname:      ${in1}.snp
indivoutname:    ${in1}.ind")
```



For more information about how to manipulate EIG, PLINK and VCF files: https://github.com/roberta-davidson/Guide-to-manipulating-PLINK-EIG-and-VCF-files
