# ADMIXTURE

## Introduction
ADMIXTURE is a software tool for maximum likelihood estimation of individual ancestries from multilocus SNP genotype datasets. 
It uses the same statistical model as STRUCTURE but calculates estimates much more rapidly using a fast numerical optimization algorithm.

Specifically, ADMIXTURE uses a block relaxation approach to alternately update allele frequency and ancestry fraction parameters. 
Each block update is handled by solving a large number of independent convex optimization problems, which are tackled using a fast sequential quadratic programming algorithm. 
Convergence of the algorithm is accelerated using a novel quasi-Newton acceleration method. The algorithm outperforms EM algorithms and MCMC sampling methods by a wide margin.

Publication describing ADMIXTURE's algorithm is
- D.H. Alexander, J. Novembre, and K. Lange. Fast model-based estimation of ancestry in unrelated individuals. Genome Research, 19:1655–1664, 2009.

The quasi-Newton convergence acceleration method is described in more detail in
- H. Zhou, D. H. Alexander, and K.  Lange. A quasi-Newton method for accelerating the convergence of iterative optimization algorithms. Statistics and Computing, 2009.

The ADMIXTURE manual says a minumum of 10,000 markers are necessary for comparing populations between continents, but at least 100,000 are better for comparing within a continent.

ADMIXTURE 1.3.0 manual: https://vcru.wisc.edu/simonlab/bioinformatics/programs/admixture/admixture-manual.pdf
Useful ADMIXTURE tutorial: https://gaworkshop.readthedocs.io/en/latest/contents/07_admixture/admixture.html 

## RUN ADMIXTURE
### Convert EIG to PLINK
```
conda activate eigensoft

in1=$1

convertf -p <(echo "genotypename:       ${in1}.geno
snpname:        ${in1}.snp
indivname:      ${in1}.ind
outputformat:   PACKEDPED
genotypeoutname:        ${in1}.bed
snpoutname:     ${in1}.bim
indivoutname:   ${in1}.fam")
```

### Prune for linkage desequilibrium
```
plink --bfile <data> --keep-allele-order --allow-no-sex --indep-pairwise 50 10 0.1 --make-bed --out <data_LDpruned>
```

### Option 1: Unsupervised ADMIXTURE
```
module load Admixture/1.3.0

BED_FILE=<path>/<input_file>.bed

for K in {2..12}; do
		admixture -C 100 -j2 -s time --cv $BED_FILE $K | tee admixture_log${K}.out
done
```
`-C 100` stops the algorithm after 100 iterations

`-j2` specifies 2 threads

`-s` time generates a random seed based on the time

`--cv` Means cross-validation will be calculated 

NOTE: At least run 10 replicates.

### Option 2: Supervised ADMIXTURE
If something is known about the relationship between populations, you can select populations as fixed source groups of ancestry and infer those ancestries in you test individuals.
Requires the flag `--supervised` and an additional `*.pop` file (with matching prefix), this has the same number of lines as the `*.fam`, one line per individual with the population names if they denote a 
fixed ancestry and those you wish to infer ancestry for have a "-" instead. For example:
```
Fam1
Fam2
-
-
-
```
Where Fam1 and Fam2 are fixed ancestries and ancestral proportion will be inferred for Ind3-Ind5.
Note ADMIXTURE will only run for K=(number of fixed ancestries), it will not infer more ancestries at higher K values than the number you specify in the `*.pop` file.
```
module load Admixture/1.3.0

in1=$1 # BED file prefix
in2=$2 # K value

admixture -C 100 -j2 -s time --supervised --cv ${in1}.bed ${in2} | tee admixture.out
```

NOTE: At least run 10 replicates.

### Plotting ADMIXTURE with PONG
Install pong (locally): https://github.com/ramachandran-lab/pong

Prepare the following files:

1. **filemap.txt**: gives paths to the matrix files that are the output of Admixture. It has 3 columns:

- `Column 1`: runID, a unique label for the Q matrix (e.g. the string “run5_K7”). Note: A runID must begin with a letter (A-Z/a-z), followed by any number of hyphens (-), underscores (_), letters, or numbers. Other characters are not allowed in runIDs. Hashmarks (#) can be used in the filemap to indicate the start of a comment.
- `Column 2`: K value for the Q matrix. Each value of K between Kmin and Kmax must be represented by at least one Q matrix in the filemap; if not, pong will abort.
- `Column 3`: Path to the Q matrix, relative to the location of the filemap. Thus, if the filemap is in the same directory as the Q matrix file, this is just the name of the Q matrix file. Note that the metadata provided in the filemap allow the user to apply pong to Q matrices in multiple directories in the user’s computer. The path cannot contain a hashmark (#) because it will be interpreted as a comment.

Example:
```
k2r1	2	data/run1/pruned_filtered_1kg_phase3.2.Q
k2r2	2	data/run2/pruned_filtered_1kg_phase3.2.Q
k2r3	2	data/run3/pruned_filtered_1kg_phase3.2.Q
k2r4	2	data/run4/pruned_filtered_1kg_phase3.2.Q
k2r5	2	data/run5/pruned_filtered_1kg_phase3.2.Q
k2r6	2	data/run6/pruned_filtered_1kg_phase3.2.Q
k2r7	2	data/run7/pruned_filtered_1kg_phase3.2.Q
k2r8	2	data/run8/pruned_filtered_1kg_phase3.2.Q
k3r1	3	data/run1/pruned_filtered_1kg_phase3.3.Q
k3r2	3	data/run2/pruned_filtered_1kg_phase3.3.Q
k3r3	3	data/run3/pruned_filtered_1kg_phase3.3.Q
k3r4	3	data/run4/pruned_filtered_1kg_phase3.3.Q
k3r5	3	data/run5/pruned_filtered_1kg_phase3.3.Q
k3r6	3	data/run6/pruned_filtered_1kg_phase3.3.Q
k3r7	3	data/run7/pruned_filtered_1kg_phase3.3.Q
k3r8	3	data/run8/pruned_filtered_1kg_phase3.3.Q
k4r1	4	data/run1/pruned_filtered_1kg_phase3.4.Q
k4r2	4	data/run2/pruned_filtered_1kg_phase3.4.Q
k4r3	4	data/run3/pruned_filtered_1kg_phase3.4.Q
k4r4	4	data/run4/pruned_filtered_1kg_phase3.4.Q
k4r5	4	data/run5/pruned_filtered_1kg_phase3.4.Q
```

2. **ind2pop.txt** file that has one population label per line, corresponding to the order of samples in the Admixture output matrices.
3. **pop_order.txt** (optional): to determine the order populations are plotted in the PONG output bar graph.

Run:
```
pong -m filemap.txt -n pop_order.txt -i ind2pop.txt
```

PONG Manual: http://brown.edu/Research/Ramachandran_Lab/files/pong/pong-manual.pdf

### Check cross-validation errors
To determine the optimal number of ancestral populations (K) in an unsupervised ADMIXTURE analysis, it's essential to examine cross-validation (CV) errors. The elbow method can then be applied to identify the K value that best captures the genetic structure of the data.

In each ADMIXTURE run, check the output file (e.g., admixture.out) for a section like this:
```
Summary: 
Converged in 100 iterations (5857.78 sec)
Loglikelihood: -164562914.531024
Fst divergences between estimated populations: 
	Pop0	Pop1	Pop2	
Pop0	
Pop1	0.144	
Pop2	0.196	0.190	
Pop3	0.273	0.241	0.289	
CV error (K=4): 0.80115
Writing output files.
```
Extract the CV error value for each replicate and each K. For example:
```
k	r	cv
3	1	0.43898
4	1	0.42526
5	1	0.42561
6	1	0.4449
7	1	0.45824
8	1	0.47691
9	1	0.4722
3	2	0.43301
4	2	0.42524
5	2	0.44515
6	2	0.44192
7	2	0.4735
8	2	0.4666
9	2	0.49127
```
Save the values in a file (e.g., cv.txt) and plot the average CV error across replicates for each K using R:

```
library(ggplot2)
library(dplyr)

cv <- read.delim("cv.txt")

#Summary tamble groupping by k 
library(dplyr)
cv.summary <- cv %>%
  group_by(k) %>%
  summarise(
    sd = sd(cv, na.rm = TRUE),
    cv = mean(cv)
  )

#Line plot with error bars
ggplot(cv.summary, aes(k, cv)) +
  geom_line(aes(group = 1)) +
  geom_errorbar( aes(ymin = cv-sd, ymax = cv+sd),width = 0.2) +
  geom_point(size = 2) +
  scale_x_continuous(breaks = 3:9)
```
Using the elbow method, identify the value of K where the CV error stops decreasing substantially. This point—often referred to as the elbow—indicates a good balance between model simplicity and explanatory power. It typically reflects the most appropriate number of ancestral populations represented in your data, helping to avoid overfitting due to unnecessarily high values of K.
