# ADMIXTURE

## Introduction
ADMIXTURE is a software tool for maximum likelihood estimation of individual ancestries from multilocus SNP genotype datasets. 
It uses the same statistical model as STRUCTURE but calculates estimates much more rapidly using a fast numerical optimization algorithm.

Specifically, ADMIXTURE uses a block relaxation approach to alternately update allele frequency and ancestry fraction parameters. 
Each block update is handled by solving a large number of independent convex optimization problems, which are tackled using a fast sequential quadratic programming algorithm. 
Convergence of the algorithm is accelerated using a novel quasi-Newton acceleration method. The algorithm outperforms EM algorithms and MCMC sampling methods by a wide margin.

Publication describing ADMIXTURE's algorithm is
- D.H. Alexander, J. Novembre, and K. Lange. Fast model-based estimation of ancestry in unrelated individuals. Genome Research, 19:1655â€“1664, 2009.

The quasi-Newton convergence acceleration method is described in more detail in
- H. Zhou, D. H. Alexander, and K.  Lange. A quasi-Newton method for accelerating the convergence of iterative optimization algorithms. Statistics and Computing, 2009.

The ADMIXTURE manual says a minumum of 10,000 markers are necessary for comparing populations between continents, but at least 100,000 are better for comparing within a continent.

ADMIXTURE 1.3.0 manual: https://vcru.wisc.edu/simonlab/bioinformatics/programs/admixture/admixture-manual.pdf
Useful ADMIXTURE tutorial: https://gaworkshop.readthedocs.io/en/latest/contents/07_admixture/admixture.html 

## Convert EIG to PLINK
```
conda activate eigensoft

in1=$1

convertf -p <(echo "genotypename:       ${in1}.geno
snpname:        ${in1}.snp
indivname:      ${in1}.ind
outputformat:   PACKEDPED
genotypeoutname:        ${in1}.bed
snpoutname:     ${in1}.bim
indivoutname:   ${in1}.fam")
```

## Prune for linkage desequilibrium
```
plink --bfile <data> --keep-allele-order --allow-no-sex --indep-pairwise 50 10 0.1 --make-bed --out <data_LDpruned>
```



