# ADMIXTURE

## Introduction
ADMIXTURE is a software tool for maximum likelihood estimation of individual ancestries from multilocus SNP genotype datasets. 
It uses the same statistical model as STRUCTURE but calculates estimates much more rapidly using a fast numerical optimization algorithm.

Specifically, ADMIXTURE uses a block relaxation approach to alternately update allele frequency and ancestry fraction parameters. 
Each block update is handled by solving a large number of independent convex optimization problems, which are tackled using a fast sequential quadratic programming algorithm. 
Convergence of the algorithm is accelerated using a novel quasi-Newton acceleration method. The algorithm outperforms EM algorithms and MCMC sampling methods by a wide margin.

Publication describing ADMIXTURE's algorithm is
- D.H. Alexander, J. Novembre, and K. Lange. Fast model-based estimation of ancestry in unrelated individuals. Genome Research, 19:1655â€“1664, 2009.

The quasi-Newton convergence acceleration method is described in more detail in
- H. Zhou, D. H. Alexander, and K.  Lange. A quasi-Newton method for accelerating the convergence of iterative optimization algorithms. Statistics and Computing, 2009.

The ADMIXTURE manual says a minumum of 10,000 markers are necessary for comparing populations between continents, but at least 100,000 are better for comparing within a continent.

ADMIXTURE 1.3.0 manual: https://vcru.wisc.edu/simonlab/bioinformatics/programs/admixture/admixture-manual.pdf
Useful ADMIXTURE tutorial: https://gaworkshop.readthedocs.io/en/latest/contents/07_admixture/admixture.html 

## RUN ADMIXTURE
### Convert EIG to PLINK
```
conda activate eigensoft

in1=$1

convertf -p <(echo "genotypename:       ${in1}.geno
snpname:        ${in1}.snp
indivname:      ${in1}.ind
outputformat:   PACKEDPED
genotypeoutname:        ${in1}.bed
snpoutname:     ${in1}.bim
indivoutname:   ${in1}.fam")
```

### Prune for linkage desequilibrium
```
plink --bfile <data> --keep-allele-order --allow-no-sex --indep-pairwise 50 10 0.1 --make-bed --out <data_LDpruned>
```

### Option 1: Unsupervised ADMIXTURE
```
module load Admixture/1.3.0

BED_FILE=<path>/<input_file>.bed

for K in {2..12}; do
		admixture -C 100 -j2 -s time --cv $BED_FILE $K | tee admixture_log${K}.out
done
```
`-C 100` stops the algorithm after 100 iterations

`-j2` specifies 2 threads

`-s` time generates a random seed based on the time

`--cv` Means cross-validation will be calculated 

### Option 2: Supervised ADMIXTURE
If something is known about the relationship between populations, you can select populations as fixed source groups of ancestry and infer those ancestries in you test individuals.
Requires the flag `--supervised` and an additional `*.pop` file (with matching prefix), this has the same number of lines as the `*.fam`, one line per individual with the population names if they denote a 
fixed ancestry and those you wish to infer ancestry for have a "-" instead. For example:
```
Fam1
Fam2
-
-
-
```
Where Fam1 and Fam2 are fixed ancestries and ancestral proportion will be inferred for Ind3-Ind5.
Note ADMIXTURE will only run for K=(number of fixed ancestries), it will not infer more ancestries at higher K values than the number you specify in the `*.pop` file.
```
module load Admixture/1.3.0

in1=$1 # BED file prefix
in2=$2 # K value

admixture -C 100 -j2 -s time --supervised --cv ${in1}.bed ${in2} | tee admixture.out
```

### Plotting ADMIXTURE with PONG
Install pong (locally):
```
conda activate
conda create -n pong 
conda install bioconda::eigensoft 
```

PONG Manual: http://brown.edu/Research/Ramachandran_Lab/files/pong/pong-manual.pdf

